app/services/creators/membership_batch_creator.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+class MembershipBatchCreator
+  # Class method to create memberships for many users with the same attributes
+  #
+  # @param [Array] user_ids The ids of users to create memberships for
+  # @param [#to_h] params The params for the memberships
+  def self.run(**params)
+    new(params).run
+  end
+
+  def initialize(user_ids:, **params)
+    @users = User.find(user_ids.reject(&:empty?))
+    @params = params
+  end
+
+  def run
+    @results = users.map do |u|
+      MembershipCreator.create!(params.to_h.merge(user: u))
+    end
+    build_result
+  end
+
+  private
+
+  attr_reader :users, :successes, :failures, :results, :params
+
+  def build_result
+    @successes, @failures = results.partition { |r| r[:msg].key? :success }
+    obj = successes.first[:object] unless successes.empty?
+    {
+      object: obj || nil,
+      msg: { notice: build_flash }
```

Comment:
This should have two keys, `:success` with the success messages and `:error` with the failures. The logic for generating it should be in `build_flash`.